### ISSUE TITLE
Geopackage - File System Watcher

### ISSUE THREAD

**Issue by Saijin-Naib** [qgis/QGIS#33690]
**Feature description.**
I'd like for there to be a file system watcher as part of the Geopackage driver that would detect when a geopackage has been updated/replaced by an external tool when loaded within QGIS, to facilitate the (re)load of data in the project seamlessly.

Inspired by Issue #33259 , and in response to comment #[565387571](https://github.com/qgis/QGIS/pull/33259#issuecomment-565387571).

**Additional context**
Would likely close/improve the below issues:

- Show Feature Count not updating properly #29616  

 - Overwrite of layer within Geopackage does not trigger updates/cleanups/maintenance as it should #29617  

 - source fields list does not update when layer is overwritten #29629

**Comment by roya0045** [comment]
@Saijin-Naib did you try with sql triggers ?

**Comment by Saijin-Naib** [comment]
> @Saijin-Naib did you try with sql triggers ?

No! I'm pretty crap with database stuff, to be completely honest. Learning as I go blindly. If I could go back in time to when I got my undergrad/grad in GIS I'd smack my advisors for not pushing DBA/SQL and Python :|

**Comment by roya0045** [comment]
If you have your layers connected in the db manager, simply go on the info page and scroll down, there should be some triggers by default. This should give you an idea of the working syntax (each provider like things differently) , [Here is some basics on trigger](https://www.tutorialspoint.com/plsql/plsql_triggers.htm) Pretty much use this in the sql window and if you can run it, it should work. 

The only things that might be an issue are geospatial functions, often they need to rely on the spatialite extension and the extension is 'stored' in the connection. This will lead to geospatial function behaving weirdly. Though something was recently implemented in the provided when I made my PR for the SQL query dialog.

**Comment by uclaros** [comment]
For non-database formats where triggers are not available, you can use `QFileSystemWatcher`
I use the following as a workaround for the broken watch file functionality on csv layers.

```
from PyQt5.QtCore import QFileSystemWatcher
watcher = QFileSystemWatcher()
watcher.addPath(filePath)
watchedLayer = vlayer
def reload_watched_layer():
    watchedLayer.dataProvider().forceReload()
watcher.fileChanged.connect(reload_watched_layer)
```
